src/Microsoft.VisualStudio.Composition/Reflection/TypeRef.cs

```diff
@@ -388,17 +420,27 @@ public bool Equals(Type other)
 
         internal void GetInputAssemblies(ISet<AssemblyName> assemblies) => ResolverExtensions.GetInputAssemblies(this, assemblies);
 
-        private static Rental<Type[]> GetResolvedTypeArray(ImmutableArray<TypeRef> typeRefs)
+        private Rental<Type[]> GetResolvedTypeArray(ImmutableArray<TypeRef> typeRefs, out bool isNullableTypeRefRequested)
         {
+            isNullableTypeRefRequested = false;
             if (typeRefs.IsDefault)
             {
+                this.DiagnosticInfoCollector.Collect($"Default type requested.");
                 return default(Rental<Type[]>);
             }
 
             var result = ArrayRental<Type>.Get(typeRefs.Length);
             for (int i = 0; i < typeRefs.Length; i++)
             {
                 result.Value[i] = typeRefs[i].ResolvedType;
+
+                this.DiagnosticInfoCollector.Collect($"arg[{i}]={result.Value[i]}");
+                if (result.Value[i] == null && !isNullableTypeRefRequested)
+                {
+                    isNullableTypeRefRequested = true;
+                    this.DiagnosticInfoCollector.Collect($"arg[{i}] is expected to be not null.");
+                    this.DiagnosticInfoCollector.Collect($"arg[{i}] Details: {typeRefs[i].DiagnosticInfoCollector.GetAndClearInformation()}");
```

Comment:
This looks like it could snowball into a _very_ large string. If you take my earlier suggestion and avoid creating strings and instead store the data as flags and collections I think you can dramatically reduce the risk of significant GC pressure being added by the change.